feat: Transfer protocol for TypeGPU resources (primarily for worklet support) (#2732)
software-mansion/TypeGPU: apps/typegpu-docs/src/content/docs/integration/react-native/worklets.mdx

---
title: React Native Worklets
description: A guide on running TypeGPU render loops on the UI thread with react-native-worklets.
---

> With [react-native-worklets](https://docs.swmansion.com/react-native-worklets/), per-frame GPU work can run on the UI thread, unaffected by a busy JS thread.

Reviewer: A minor suggestion
```suggestion
With [react-native-worklets](https://docs.swmansion.com/react-native-worklets/), per-frame GPU work can be scheduled on the UI thread, unaffected by tasks running on the default React Native thread (or RN thread).
```

> TypeGPU resources created on the JS thread can be captured by worklets directly, they are transferred between runtimes automatically.

Reviewer: ```suggestion
TypeGPU resources created on the RN thread can be captured by worklets directly, they are transferred between runtimes automatically.
```

## Setup

Follow the [React Native guide](/TypeGPU/integration/react-native/) first, then install `react-native-worklets` and enable its babel plugin with [Bundle Mode](https://docs.swmansion.com/react-native-worklets/docs/bundleMode/):

```diff lang=js title="babel.config.js"
+const workletsPluginOptions = {
+  bundleMode: true,
+  importForwarding: {
+    moduleNames: ['typegpu'],
+    // Directories with your module-scope shader definitions
+    relativePaths: ['my-app/components'],
+  },
+};

module.exports = (api) => {
  api.cache(true);
  return {
    presets: ['babel-preset-expo'],
    plugins: [
      'unplugin-typegpu/babel',
+      ['react-native-worklets/plugin', workletsPluginOptions],
    ],
  };
};
```

Import the hooks from the dedicated entrypoint - importing it also registers the transfer support for TypeGPU resources:

```ts
import { useRoot, useFrame, useConfigureContext, useUniform } from '@typegpu/react/react-native-worklets';
```

After changing the babel config, clear the Metro cache with `npx expo start --clear`.

## Example

> Create resources on the JS thread, then use them freely inside a `useFrame` worklet:

Reviewer: ```suggestion
Create resources on the RN thread, then use them freely inside a `useFrame` worklet:
```

      root.createRenderPipeline({
        vertex: common.fullScreenTriangle,
        fragment: () => {
          'use gpu';
          return d.vec4f(color.$, 1);
        },
      }),
    [root, color],
  );

  const { ref, ctxRef } = useConfigureContext({ alphaMode: 'premultiplied' });

  // Runs each frame on the UI thread
  useFrame(({ elapsedSeconds }) => {
    'worklet';
    const ctx = ctxRef.value;
    if (!ctx) return;

    color.write(d.vec3f(0.5 + Math.sin(elapsedSeconds) * 0.5, 0.447, 0.941));
    pipeline.withColorAttachment({ view: ctx }).draw(3);
    ctx.present?.();
  });

  return <Canvas ref={ref} style={{ aspectRatio: 1 }} transparent />;
}
```

The `color` uniform and `pipeline` captured by the worklet are transferred to the UI runtime on first use.
Both runtimes share the same underlying GPU objects, and transferring the same resource again yields the same object back.

This works for buffers (including `createUniform`/`createMutable`/`createReadonly`), textures, samplers, bind groups and their layouts, vertex layouts, query sets, pipelines, roots, slots, accessors, consts, and vector/matrix instances.

## Rules of transfer

**Definitions are runtime-local.**

> Shader functions (`tgpu.fn`, entry functions), `tgpu.comptime`, and schemas cannot be serialized.

Reviewer 2: Schemas cannot be serialized?

  const pipeline = useMemo(
    () =>
      root.createRenderPipeline({
        vertex: common.fullScreenTriangle,
        fragment: () => {
          'use gpu';
          return d.vec4f(color.$, 1);
        },
      }),
    [root, color],
  );

  const { ref, ctxRef } = useConfigureContext({ alphaMode: 'premultiplied' });

  // Runs each frame on the UI thread
  useFrame(({ elapsedSeconds }) => {
    'worklet';
    const ctx = ctxRef.current;
    if (!ctx) return;

    color.write(d.vec3f(0.5 + Math.sin(elapsedSeconds) * 0.5, 0.447, 0.941));
    pipeline.withColorAttachment({ view: ctx }).draw(3);
    ctx.present?.();
  });

  return <Canvas ref={ref} style={{ aspectRatio: 1 }} transparent />;
}
```

The `color` uniform and `pipeline` captured by the worklet are transferred to the UI runtime on first use.
Both runtimes share the same underlying GPU objects, and transferring the same resource again yields the same object back.

This works for buffers (including `createUniform`/`createMutable`/`createReadonly`), textures, samplers, bind groups and their layouts, vertex layouts, query sets, pipelines, roots, slots, accessors, consts, and vector/matrix instances.

## Opting out

> To keep everything on the JS thread even with `react-native-worklets` installed, pass `disableWorklets` to the `Root` provider:

Reviewer: ```suggestion
To keep everything on the RN thread even with `react-native-worklets` installed, pass `disableWorklets` to the `Root` provider:
```

```tsx
<Root disableWorklets>
  <Pulse />
</Root>
```

> `useFrame` then runs its callbacks on the JS thread, and `useConfigureContext` returns a plain object of the same shape, so the code above keeps working unchanged.

Reviewer: ```suggestion
`useFrame` then runs its callbacks on the RN thread, and `useConfigureContext` returns a plain object of the same shape, so the code above keeps working unchanged.
```
